Arm/State.lean

```diff
@@ -662,3 +662,318 @@ theorem write_mem_irrelevant :
+
+/--
+This is a low level theorem.
+Prefer using theorems from `Arm.Memory.Separate` that provide higher level theorems
+in terms of memory (non)-interference.
+-/
+theorem write_mem'_of_neq (hix : ix ≠ addr) : write_mem' addr val s ix = s ix := by
+  simp only [write_mem']
+  apply store_read_over_write_different
+  assumption
+
+theorem write_mem_eq_write_mem' :  (write_mem addr val s).mem = write_mem' addr val s.mem := rfl
+
+def write_mem_bytes' (n : Nat) (addr : BitVec 64)
+    (val : BitVec (n * 8)) (s : Memory) : Memory :=
+  match n with
+  | 0 => s
+  | n' + 1 =>
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n' * 8) (val >>> 8)
+    write_mem_bytes' n' (addr + 1#64) val_rest s
+
+@[simp]
+theorem write_mem_bytes_eq_write_mem_bytes' (s : ArmState) :
+    write_mem_bytes n addr val s =
+    { s with mem := write_mem_bytes' n addr val s.mem } := by
+  induction n generalizing addr s
+  case zero => simp [write_mem_bytes, write_mem_bytes']
+  case succ n' ih =>
+    simp [write_mem_bytes, write_mem_bytes', read_mem_eq_read_mem', ih,
+      write_mem, write_mem']
+
+/-- Writing zero bytes does not change memory. -/
+theorem write_mem_bytes'_zero : write_mem_bytes' 0 addr val s = s := rfl
+
+/-- Writing (n + 1) bytes can be described as writing `n` bytes and then recursing to write the rest. -/
+theorem write_mem_bytes'_succ :
+    write_mem_bytes' (n + 1) addr val s =
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n * 8) (val >>> 8) -- TODO: rewrite this as 'truncate'.
+    write_mem_bytes' n (addr + 1#64) val_rest s := rfl
+
+theorem write_mem_bytes'_eq_of_le {ix base : BitVec 64}
+    (hix : ix.toNat < base.toNat) (hnowrap : base.toNat + n ≤ 2^64) :
+    write_mem_bytes' n base data mem ix = mem ix := by
+  induction n generalizing base mem ix
+  case zero => simp [write_mem_bytes']
+  case succ n ih =>
+    simp only [write_mem_bytes']
+    rcases n with rfl | n
+    · rw [write_mem_bytes'_zero]
+      apply write_mem'_of_neq (BitVec.neq_of_lt hix)
+    · rw [ih]
+      · apply write_mem'_of_neq (BitVec.neq_of_lt hix)
+      · rw [BitVec.toNat_add_eq_toNat_add_toNat]
+        · omega
+        · simp; omega
+      · rw [BitVec.toNat_add_eq_toNat_add_toNat (by simp; omega)]
+        simp; omega
+
+theorem write_mem_bytes'_eq_of_ge {ix base : BitVec 64}
+    (hix : ix.toNat ≥ base.toNat + n)
+    (hnowrap : base.toNat + n ≤ 2^64) :
+    write_mem_bytes' n base data mem ix = mem ix := by
+  induction n generalizing base mem ix
+  case zero => simp [write_mem_bytes']
+  case succ n ih =>
+    simp only [write_mem_bytes']
+    rw [ih]
+    · have hix : ix.toNat > base.toNat := by omega
+      obtain hix : ix.toNat ≠ base.toNat := by omega
+      apply write_mem'_of_neq (by apply BitVec.neq_of_toNat_neq hix)
+    · rw [BitVec.toNat_add_eq_toNat_add_toNat (by simp; omega)]
+      simp; omega
```

Comment:
`simp only...` please, and please check in other spots too.

Reply:
We allow `simp; omega`, since they are both closing tactics. This is [standard in Lean upstream](https://github.com/leanprover/lean4/blob/11be29e68c8dea408f6bc74033d14181d015a526/src/Init/Data/BitVec/Bitblast.lean#L265). I've checked that `simp`'s that are not followed by `omega` are `simp only`.

Reply:
Yes, but in order to be more robust to the changes in general simp sets upstream, we prefer to use `only` in LNSym as much as possible.